agent/main/lib/WebsocketSession.ts

```diff
@@ -0,0 +1,183 @@
+
+const SCRIPT_PLACEHOLDER = '';
+const { log } = Log(module);
+
+export class WebsocketSession extends TypedEventEmitter<IWebsocketEvents> {
+  readonly isReady: Promise<void>;
+
+  private readonly host = 'websocket.localhost';
+  private port: number;
+  private readonly secret = Math.random().toString();
+
+  // We store resolvable when we received websocket message before, receiving
+  // targetId, this way we can await this, and still trigger to get proper ids.
+  private clientIdToTargetId = new Map<string, Resolvable<string> | string>();
+
+  private server: Server;
+  private wss: WebsocketServer;
+
+  constructor() {
+    super();
+    this.server = createServer();
+    this.wss = new WebsocketServer({ noServer: true });
+  }
+
+  async initialize(): Promise<void> {
+    const resolver = new Resolvable<void>(10e3);
+
+    this.server.on('error', resolver.reject);
+    this.server.listen(0, () => {
+      const address = this.server.address();
+      if (typeof address === 'string') {
+        throw new Error('Unexpected server address format (string)');
+      }
+      this.port = address.port;
+      resolver.resolve();
+    });
+
+    this.server.on('upgrade', this.handleUpgrade.bind(this));
+    this.wss.on('connection', this.handleConnection.bind(this));
+
+    return resolver.promise;
+  }
+
+  close(): void {
+    this.wss.close();
+    this.server.close();
+  }
+
+  isWebsocketUrl(url: string): boolean {
+    const parsed = new URL(url);
+    return (
+      parsed.hostname === this.host &&
+      parsed.port === this.port.toString() &&
+      parsed.searchParams.get('secret') === this.secret
+    );
+  }
+
+  registerWebsocketFrameId(url: string, frameId: string): void {
+    const parsed = new URL(url);
+    if (parsed.searchParams.get('secret') !== this.secret) return;
+    const clientId = parsed.searchParams.get('clientId');
+    if (!clientId) return;
+
+    const targetId = this.clientIdToTargetId.get(clientId);
+    if (targetId instanceof Resolvable) {
+      targetId.resolve(frameId);
+    }
+    this.clientIdToTargetId.set(clientId, frameId);
+  }
+
+  injectWebsocketCallbackIntoScript(script: string): string {
+    // We could do this as a simple template script but this logic might get
+    // complex over time and we want typescript to be able to check proxyScript();
+    const scriptBody = this.proxyScript
+      .toString()
+      // eslint-disable-next-line no-template-curly-in-string
+      .replaceAll('${this.host}', this.host)
+      // eslint-disable-next-line no-template-curly-in-string
+      .replaceAll('${this.port}', this.port.toString())
+      // eslint-disable-next-line no-template-curly-in-string
+      .replaceAll('${this.secret}', this.secret)
+      // Use function otherwise replace will try todo some magic
+      .replace('SCRIPT_PLACEHOLDER', () => script);
+
+    const wsScript = `(function ${scriptBody})();`;
+    return wsScript;
+  }
+
+  private proxyScript(): void {
```

Comment:
This threw me off.  I think we need to name this with something about injected, and add a comment that explains this is actually just a wrapper function to generate the string of the function. 

I also wonder if it shouldn't be a private function - instead maybe a function in the same module, just to be super clear this is not a normal function